ext/bcmath/bcmath.c

```diff
@@ -777,4 +791,938 @@ PHP_FUNCTION(bcscale)
+	intern->num = ret;
+	intern->scale = scale;
+	return intern;
+}
+
+static zend_always_inline zend_result bcmath_number_parse_num(zval *zv, zend_object **obj, zend_string **str, zend_long *lval)
+{
+	if (Z_TYPE_P(zv) == IS_OBJECT && instanceof_function(Z_OBJCE_P(zv), bcmath_number_ce)) {
+		*obj = Z_OBJ_P(zv);
+		return SUCCESS;
+	} else {
+		switch (Z_TYPE_P(zv)) {
+			case IS_LONG:
+				*lval = Z_LVAL_P(zv);
+				return SUCCESS;
+
+			case IS_STRING:
+				*str = Z_STR_P(zv);
+				return SUCCESS;
+
+			default:
+				return zend_parse_arg_str_or_long_slow(zv, str, lval, 1 /* dummy */) ? SUCCESS : FAILURE;
+		}
+	}
+}
+
+static zend_always_inline void bc_num_from_zend_object(const zend_object *obj, bc_num *num, size_t *full_scale)
+{
+	bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+	*num = intern->num;
+	if (full_scale) {
+		*full_scale = intern->scale;
+	}
+}
+
+static zend_always_inline zend_result bc_num_from_zend_string(const zend_string *str, bc_num *num, size_t *full_scale)
+{
+	if (UNEXPECTED(php_str2num_ex(num, str, full_scale) == FAILURE)) {
+		bc_free_num(num);
+		return FAILURE;
+	}
+	return SUCCESS;
+}
+
+static zend_always_inline void bc_num_from_zend_long(zend_long lval, bc_num *num, size_t *full_scale)
+{
+	php_long2num(num, lval);
+	if (full_scale) {
+		*full_scale = 0;
+	}
+}
+
+static zend_always_inline zend_result bc_num_from_obj_or_str_or_long(
+	bc_num *num, size_t *full_scale, const zend_object *obj, const zend_string *str, zend_long lval)
+{
+	if (obj) {
+		bc_num_from_zend_object(obj, num, full_scale);
+		return SUCCESS;
+	} else if (str) {
+		return bc_num_from_zend_string(str, num, full_scale);
+	} else {
+		bc_num_from_zend_long(lval, num, full_scale);
+		return SUCCESS;
+	}
+}
+
+static zend_result bcmath_number_do_operation(uint8_t opcode, zval *ret_val, zval *op1, zval *op2)
+{
+	switch (opcode) {
+		case ZEND_ADD:
+		case ZEND_SUB:
+		case ZEND_MUL:
+		case ZEND_DIV:
+		case ZEND_MOD:
+		case ZEND_POW:
+			break;
+		default:
+			return FAILURE;
+	}
+
+	zend_object *obj1 = NULL;
+	zend_string *str1 = NULL;
+	zend_long lval1;
+
+	zend_object *obj2 = NULL;
+	zend_string *str2 = NULL;
+	zend_long lval2;
+
+	if (UNEXPECTED(Z_TYPE_P(op1) == IS_NULL)) {
+		lval1 = 0;
+	} else if (UNEXPECTED(bcmath_number_parse_num(op1, &obj1, &str1, &lval1) == FAILURE)) {
+		return FAILURE;
+	}
+
+	if (UNEXPECTED(Z_TYPE_P(op2) == IS_NULL)) {
+		lval2 = 0;
+	} else if (UNEXPECTED(bcmath_number_parse_num(op2, &obj2, &str2, &lval2) == FAILURE)) {
+		return FAILURE;
+	}
+
+	bc_num n1 = NULL;
+	bc_num n2 = NULL;
+	size_t n1_full_scale;
+	size_t n2_full_scale;
+	if (UNEXPECTED(bc_num_from_obj_or_str_or_long(&n1, &n1_full_scale, obj1, str1, lval1) == FAILURE ||
+		bc_num_from_obj_or_str_or_long(&n2, &n2_full_scale, obj2, str2, lval2) == FAILURE)) {
+		zend_value_error("Number is not well-formed");
+		goto fail;
+	}
+
+	/* For increment and decrement, etc */
+	zval op1_copy;
```

Comment:
I'm confused why you need this

Reply:
GMP does the same thing, if the left operand and result have identical zvals, then modifying result also modifies the left operand.

Since this is an immutable class, this process ensures that the original object is not changed.

Reply:
Fixed in https://github.com/php/php-src/pull/13741/commits/d7483f8d17b42eb9937c2175bf986c575365de0a